File: 2018/rust/src/sleepingguards.rs
use regex::Regex;
use std::collections::HashMap;

fn solve1(sleep_detail: &HashMap<u32, HashMap<u32, u32>>) -> u32 {

    let sleepiest_guard: &u32 = sleep_detail
        .keys()
        .max_by_key(|k| sleep_detail.get(k).unwrap().values().sum::<u32>())
        .unwrap();
    let max_hours_slept: &u32 = sleep_detail
        .get(sleepiest_guard)
        .unwrap()
        .keys()
        .max_by_key(|k| sleep_detail.get(sleepiest_guard).unwrap().get(k).unwrap())
        .unwrap();
    (*sleepiest_guard) * (*max_hours_slept)
}

fn solve2(sleep_detail: &HashMap<u32, HashMap<u32, u32>>) -> u32 {
    let sleepiest_guard: &u32 = sleep_detail
        .keys()
        .max_by_key(|k| sleep_detail.get(k).unwrap().values().max())
        .unwrap();
    let sleepiest_minute: &u32 = sleep_detail
        .get(sleepiest_guard)
        .unwrap()
        .keys()
        .max_by_key(|k| sleep_detail.get(sleepiest_guard).unwrap().get(k).unwrap())
        .unwrap();
    (*sleepiest_guard) * (*sleepiest_minute)
}

pub fn solve(data: String) {
    let pattern: Regex = Regex::new(r"\d+").unwrap();

    let mut sleep_detail: HashMap<u32, HashMap<u32, u32>> = HashMap::new();
    let mut current_guard: u32 = 0;
    let mut start: u32 = 0;

    let mut input: Vec<&str> = data.split("\n")
        .filter(|l| l.len() > 0)
        .collect();
    input.sort();

    for line in input {
        let numbers: Vec<u32> = pattern
            .find_iter(&line)
            .into_iter()
            .map(|m| m.as_str().parse::<u32>().unwrap())
            .collect();

        let suffix = line.get(line.len() - 2..).unwrap();

        if suffix == "ft" {
            current_guard = *numbers.last().unwrap();
        } else if suffix == "ep" {
            start = numbers[4];
        } else if suffix == "up" {
            for hour in start .. numbers[4] {
                let entry = sleep_detail
                    .entry(current_guard)
                    .or_insert(HashMap::new())
                    .entry(hour)
                    .or_insert(0);
                *entry += 1;
            }
        } else {
            panic!("invalid input");
        }
    }

    println!("Part 1: {}", solve1(&sleep_detail));
    println!("Part 2: {}", solve2(&sleep_detail));
}

File: 2018/haskell/src/SleepingGuards.hs
module SleepingGuards (
  sleepingGuards,
  solve1,
  solve2,
  prepareInput
  ) where

import AdventOfCode (cardinality, maxKey)

import Data.Time (UTCTime, readSTime, defaultTimeLocale, LocalTime, todMin, localTimeOfDay)
import Data.List (sort)

type Guard = Int
data Action = Start Guard | Wake | Sleep
  deriving (Show, Read, Eq, Ord)

parseAction :: String -> Action
parseAction str = case words str of
  ["wakes", "up"]     -> Wake
  ["falls", "asleep"] -> Sleep
  ["Guard", '#':guardid, "begins", "shift"] -> Start (read guardid)
  _                   -> error ("error parsing action: " ++ str)

parseLine :: String -> (LocalTime, Action)
parseLine str = case readSTime True defaultTimeLocale "[%Y-%m-%d %H:%M] " str of
  [(time, remaining)] -> (time, parseAction remaining)
  _                   -> error ("error parsing line: " ++ str)

toSleepMinutes :: [(LocalTime, Action)] -> [(Guard, Int)]
toSleepMinutes = clean . expand . go (error "no start")
  where
    clean     = filter (\(guardid, _) -> guardid > 0)
    getMinute = todMin . localTimeOfDay
    expand xs = [(guardid, i)
                | (guardid, (start, end)) <- xs
                , i <- [getMinute start .. getMinute end - 1]
                ]
    go _ ((_, Start guardid) : xs)                 = go guardid xs
    go guardid ((start, Sleep) : (end, Wake) : xs) = (guardid, (start, end)) : go guardid xs
    go _ []                                        = []
    go _ xs                                        = error "invalid sequence of events"

solve1 :: [(Guard, Int)] -> Int
solve1 sleepDetail = sleepiestGuard * maxHoursSlept
  where sleepiestGuard = maxKey (cardinality [n | (n, _) <- sleepDetail])
        maxHoursSlept  = maxKey (cardinality [m | (n,m) <- sleepDetail, n == sleepiestGuard])

solve2 :: [(Guard, Int)] -> Int
solve2 sleepDetail = sleepiestGuard * sleepiestHour
  where (sleepiestGuard, sleepiestHour) = maxKey (cardinality sleepDetail)

-- test helper
prepareInput :: [String] -> [(Guard, Int)]
prepareInput = toSleepMinutes . (map parseLine) . sort

sleepingGuards :: [String] -> IO ()
sleepingGuards input = do
  let sortedInput = sort input
  let sleepDetail = toSleepMinutes (map parseLine sortedInput)
  print $ solve1 sleepDetail
  print $ solve2 sleepDetail
